packages/control/bat_all.py

```diff
@@ -272,3 +291,53 @@ def power_for_bat_charging(self):
         except Exception:
             log.exception("Fehler im Bat-Modul")
             return 0
+
+    def set_power_limit_controlable(self):
+        controlable = 0
+        for bat_component in get_controlable_bat_components():
+            if "set_power_limit" in type(bat_component).__dict__:
+                controlable += 1
+        self.data.get.power_limit_controlable = controlable
+
+    def get_power_limit(self):
+        if (self.data.config.power_limit_mode != BatPowerLimitMode.NO_LIMIT.value
+                and len(get_chargepoints_by_chargemodes(self.CHARGEMODES_FOR_POWER_LIMIT)) > 0 and
+                self.data.get.power_limit_controlable > 0 and
+                # Nur wenn kein Überschuss im System ist, Speicherleistung begrenzen.
+                self.data.get.power <= 0 and
+                data.data.counter_all_data.get_evu_counter().data.get.power >= 0):
+            if self.data.config.power_limit_mode == BatPowerLimitMode.LIMIT_STOP.value:
+                self.data.set.power_limit = 0
+            elif self.data.config.power_limit_mode == BatPowerLimitMode.LIMIT_TO_HOME_CONSUMPTION.value:
+                self.data.set.power_limit = data.data.counter_all_data.data.set.home_consumption
+            log.debug(f"Speicher-Leistung begrenzen auf {self.data.set.power_limit/1000}kW "
+                      f"({self.data.get.power_limit_controlable} Speicher)")
+        else:
+            self.data.set.power_limit = None
+            if len(get_chargepoints_by_chargemodes(self.CHARGEMODES_FOR_POWER_LIMIT)) == 0:
+                log.debug("Speicher-Leistung nicht begrenzen, "
+                          "da keine Ladepunkte in einem Lademodus mit Netzbezug sind.")
+            elif self.data.get.power_limit_controlable == 0:
+                log.debug("Speicher-Leistung nicht begrenzen, da keine regelbaren Speicher vorhanden sind.")
+            elif self.data.get.power > 0:
+                log.debug("Speicher-Leistung nicht begrenzen, da kein Speicher entladen wird.")
+            elif data.data.counter_all_data.get_evu_counter().data.get.power < 0:
+                log.debug("Speicher-Leistung nicht begrenzen, da EVU-Überschuss vorhanden ist.")
+            else:
+                log.debug("Speicher-Leistung nicht begrenzen.")
+        for bat_component in get_controlable_bat_components():
+            if self.data.set.power_limit is None:
+                power_limit = None
+            else:
+                power_limit = self.data.set.power_limit / self.data.get.power_limit_controlable
+            data.data.bat_data[f"bat{bat_component.component_config.id}"].data.set.power_limit = power_limit
```

Comment:
Wenn ich das richtig verstanden habe, wird die gesamte Leistungsgrenze (`power_limit`) gleichmäßig auf die steuerbaren Speicher (`power_limit_controlable`) aufgeteilt. Dabei wird jedoch nicht beachtet, ob der jeweilige Speicher diese Leistung überhaupt bereitstellen kann.

Einzelne Speicher könnten bereits leer sein oder die Leistung wird durch den WR (direkt vom Speicher oder vom Hybrid-WR) begrenzt. In diesem Fall sollten andere Speicher eine höhere Leistung zugeteilt bekommen, um nicht gewollten Netzbezug zu verhindern.

Reply:
umgesetzt